storage/executable-store/src/lib.rs

```diff
@@ -0,0 +1,588 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_infallible::Mutex;
+use aptos_logger::{error, info};
+use aptos_types::executable::Executable;
+use dashmap::DashMap;
+use std::{
+    fmt::Debug,
+    hash::Hash,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+/// Represents the state of the ExecutableStore. Introduced to ensure the proper usage
+/// of the ExecutableStore throughout the system, as due to the current codebase structure
+/// constraints, different important steps are performed in different crates. Such
+/// defensive programming here ensures correctness (clearing the cache on any unexpected
+/// state updates) and observability to quickly identify any incorrect usage.
+///
+/// Initial state is Empty, and it is also possible to get to the Empty state by clearing
+/// the ExecutableStore at any time.
+///
+/// Execution in the system happens piece-wiese, where each piece contains a continuous
+/// sub-sequence of ledger of the ordered transactions. For example, in our codebase,
+/// pieces are called 'blocks' during the original execution / 'chunks' during state-sync.
+/// The only exception to this occurs while the cache in the Before(id) state is updated,
+/// adding and removing elements (potentially concurrently) in order to convert it to
+/// contain the executables at version corresponding to after the execution of the piece.
+/// When the update is done, it is the user's responsibility to mark the state as Updated.
+/// Currently, this happens after parallel execution in the block-executor crate.
+///
+/// After updating the ExecutableStore, it must be pruned to control the memory usage.
+/// Currently, this happens in aptos-vm based on a single threshold, clearing the store
+/// if the size exceeds the threshold. Then, it is the caller's responsibility to change
+/// the status to Pruned (ensuring we will never use the ExecutableStore without pruning).
+/// After pruning, the state may change to After(id), with the id of the latest piece.
+/// To go back to the Before(id') state, the parent of piece id' is conformed to be id,
+/// ensuring correctness. This currently occurs in execution/executor, as the information
+/// about pieces (ids and their relation to each other) is only available there.
+///
+/// The full state transition diagram is below:
+///
+/// Empty ----------> Updated(None)
+///                      |
+///                      |
+///                      ↓
+///                   Pruned(None)
+///                      |
+///                      |
+///                      ↓
+///   --------------- After(id) -------------> Before(id')
```

Comment:
Why cannot we merge Pruned and After?

Reply:
Pruning happens at aptos-vm, which is the correct place for the eventual correct architecture of these things.
After is is set at outside executors, as that's where we know IDs.

Also they serve different purpose, one protects against not pruning, the other, in combination with before, protects against running with the wrong cache with respect to the (previous/current) block